api/src/sparql.rs

```diff
@@ -0,0 +1,102 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//!
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::graph::Graph;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset {
+    type BindingsTerm: TTerm;
+    type BindingsResult: SparqlBindings<Self>;
+    type GraphResult: Graph;
+    type SparqlError: Error + 'static;
+
+    /// Parse and immediately execute `query`
+    fn query(&self, query: &str) -> Result<SparqlResult<Self>, Self::SparqlError>;
+}
+
+/// The result of executing a SPARQL query.
+pub enum SparqlResult<T>
+where
+    T: SparqlDataset + ?Sized,
+{
+    Bindings(T::BindingsResult),
+    Boolean(bool),
+    Graph(T::GraphResult),
+}
+
+/// The result of executing a SPARQL SELECT query
+pub trait SparqlBindings<D>:
+    IntoIterator<Item = Result<Vec<Option<D::BindingsTerm>>, D::SparqlError>>
+where
+    D: SparqlDataset + ?Sized,
+{
+    /// Return the list of SELECTed variable names
+    fn variables(&self) -> Result<Vec<String>, D::SparqlError>;
```

Comment:
I don't know if there is a usecase where this method could fail (it does not in Oxigraph).

Reply:
If we assume that some implementations do not own the variable names, then we must provision for the situation were such implementations would fail to retrieve the variable names when required to. Some may e.g. perform some IO in the method `variables()`, and this IO may fail...

That being said, the more I think about it, the more I find this objection of mine rather theoretical... Consider [the XML format for SPARQL endpoint responses](https://www.w3.org/TR/rdf-sparql-XMLres/). In order to know whether a response is a Bindings or a Boolean, implementations have to go past the `<head>` element, hence to parse the variable names. So any failure there will caught by the `query` method...

So maybe the assumption that any impl of `SparqlBindings` owns a copy of the variable names, is not so bad after all... WDYT?

Reply:
I think ownership and certainty about the availability of the data are two separate concerns. The client could have very certainly available data regardless of whether it is owned or borrowed.

I think it should be part of the contract that if you own a `SparqlResult` type, then the results are necessarily available. Otherwise, the query would have been considered a failure (so, either IO issues or parsing issues with the result would fail in the query itself). 

I also think it doesn't really make sense to expect IO in the `SparqlResult` implementation. If you're trying to provide an API for lazy loading the results or streaming, then this could perhaps be created by a different query function, maybe `try_query()` which would return a `Result` of `SparqlLazyResult`. I'm not sure how much of a real use-case this is.

This API has now 2 levels of indirection: starting from `query()` you have a `Result` which can be `Err` or `Ok`, and then `SparqlResult` which can be `Bindings`, `Boolean` or `Triples`. Adding another `Result` to the mix seems to me a bit too defensive.

I like this API much better than the previous one (#90) and find it good. I still think it might make sense to create an API (maybe an additional function?) that allows the user to specify the expect query form and not have to destructure the result.

These are 2 different use-cases: 
1. the client code could be getting a SPARQL query from some input (say, if it's a server that gets a SPARQL query as an http request payload without knowing what query form it gets)
2. the client code could hard code a specific query, knowing in advance the query form and probably also the variables used, etc., if it's a SELECT query. Even if it's not hard-coded, the client code could expect a specific query form.

For the second use case, I think it would be helpful to have an API for doing:
```rust
let bindings = dataset.make_specific_query::<SparqlSelectQuery>("SELECT ...")?;
```

or if you don't like generics, something like:
```rust
let bindings = dataset.select_query("SELECT ...")?;
let boolean = dataset.ask_query("ASK ...")?;
let triples = dataset.construct_query("CONSTRUCT ...")?;
```

Reply:
@yever 
> I think ownership and certainty about the availability of the data are two separate concerns.

You are right. Talking about ownership was a shortcut for "being able to guarantee that the `&str` can leave long enough".

> I also think it doesn't really make sense to expect IO in the `SparqlResult` implementation.

I disagree. A `SparqlResult` implementation might be streaming on the response of a SPARQL endpoint. Another implementation (such as Oxigraph) might be streaming results from a disk-based dataset, that might fail at anypoint. 

That being said, and since this was the starting point of this comment, I am less and less convinced that `variables` should return a `Result`.

>  I still think it might make sense to create an API (maybe an additional function?) that allows the user to specify the expected query form and not have to destructure the result.

This is a different issue. See below (after commit 34a3db7).

Reply:
> > I also think it doesn't really make sense to expect IO in the `SparqlResult` implementation.
> 
> I disagree. A `SparqlResult` implementation might be streaming on the response of a SPARQL endpoint. Another implementation (such as Oxigraph) might be streaming results from a disk-based dataset, that might fail at anypoint.
> 
> That being said, and since this was the starting point of this comment, I am less and less convinced that `variables` should return a `Result`.

As discussed in the other thread (https://github.com/pchampin/sophia_rs/pull/91#discussion_r478484044), I think this is a question of whether the API is generic and accounts for both fallible and infallible implementations, or should the APIs be split for the two cases.

I think it's fine to have the former option. It just means that a user of an infallible implementation will have to `unwrap()` a bit.

Reply:
@pchampin 

+1 to make `variables()` infallible. Parsers should be able to parse the file beginning before allowing calls to this method. Oxigraph has implemented a XML query results parser without problem doing so.

> So maybe the assumption that any impl of SparqlBindings owns a copy of the variable names, is not so bad after all... WDYT?

Indeed. For example, in Oxigraph, I have to wrap the`QuerySolutionIter` type to make it implement `SparqlBindings` so I could easily add a `variables: Vec<String>` struct member. I guess this could also be easily done by the other implementation (result parsers...).